flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/RemoteStreamEnvironment.java

```diff
@@ -171,116 +142,82 @@ public RemoteStreamEnvironment(String host, int port, Configuration clientConfig
+	}
 
-		Configuration configuration = new Configuration();
-		configuration.addAll(clientConfiguration);
+	private static List<URL> getClasspathURLs(final URL[] classpaths) {
+		return classpaths == null ? Collections.emptyList() : Arrays.asList(classpaths);
+	}
 
-		configuration.setString(JobManagerOptions.ADDRESS, host);
-		configuration.setInteger(JobManagerOptions.PORT, port);
+	private static Configuration validateAndGetEffectiveConfiguration(
+			final Configuration configuration,
+			final String host,
+			final int port,
+			final String[] jarFiles,
+			final URL[] classpaths,
+			final SavepointRestoreSettings savepointRestoreSettings) {
+		RemoteEnvironmentConfigUtils.validate(host, port);
+		return getEffectiveConfiguration(
+				getClientConfiguration(configuration),
+				host,
+				port,
+				jarFiles,
+				getClasspathURLs(classpaths),
+				savepointRestoreSettings);
+	}
 
-		configuration.setInteger(RestOptions.PORT, port);
+	private static Configuration getEffectiveConfiguration(
+			final Configuration baseConfiguration,
+			final String host,
+			final int port,
+			final String[] jars,
+			final List<URL> classpaths,
+			final SavepointRestoreSettings savepointRestoreSettings) {
 
-		final ClusterClient<?> client;
-		try {
-			client = new RestClusterClient<>(configuration, "RemoteStreamEnvironment");
-		}
-		catch (Exception e) {
-			throw new ProgramInvocationException("Cannot establish connection to JobManager: " + e.getMessage(), e);
-		}
+		final Configuration effectiveConfiguration = new Configuration(baseConfiguration);
+
+		RemoteEnvironmentConfigUtils.setJobManagerAddressToConfig(host, port, effectiveConfiguration);
+		RemoteEnvironmentConfigUtils.setJarURLsToConfig(jars, effectiveConfiguration);
+		ConfigUtils.encodeCollectionToConfig(effectiveConfiguration, PipelineOptions.CLASSPATHS, classpaths, URL::toString);
 
 		if (savepointRestoreSettings != null) {
-			streamGraph.setSavepointRestoreSettings(savepointRestoreSettings);
+			SavepointRestoreSettings.toConfiguration(savepointRestoreSettings, effectiveConfiguration);
+		} else {
+			SavepointRestoreSettings.toConfiguration(SavepointRestoreSettings.none(), effectiveConfiguration);
 		}
 
-		try {
-			final PlanExecutor executor = PlanExecutor.createRemoteExecutor(
-					host,
-					port,
-					clientConfiguration);
+		// these should be set in the end to overwrite any values from the client config provided in the constructor.
+		effectiveConfiguration.setString(DeploymentOptions.TARGET, "remote-cluster");
+		effectiveConfiguration.setBoolean(DeploymentOptions.ATTACHED, true);
+
+		return effectiveConfiguration;
+	}
 
-			return executor.executePlan(streamGraph, jarFiles, globalClasspaths).getJobExecutionResult();
+	@Override
+	public JobExecutionResult execute(StreamGraph streamGraph) throws Exception {
+		transformations.clear();
+		try {
+			return super.execute(streamGraph);
 		}
 		catch (ProgramInvocationException e) {
```

Comment:
Maybe we just

```java
return super.execute(steamGraph);
```

which keep the same as batch codepath. The concern here is that `ProgramInvocationException` is inside flink-client and in the near further we hopefully reverse the dependency from flink-streaming-java to flink-client.